operator/pkg/values/map.go

```diff
@@ -0,0 +1,517 @@
+func ConvertMap[T any](m Map) (T, error) {
+	return fromJSON[T]([]byte(m.JSON()))
+}
+
+func setPathRecurse(base map[string]any, paths []string, value any) error {
+	seg := paths[0]
+	last := len(paths) == 1
+	nextIsArray := len(paths) >= 2 && strings.HasPrefix(paths[1], "[")
+	if nextIsArray {
+		last = len(paths) == 2
+		// Find or create target list
+		if _, f := base[seg]; !f {
+			base[seg] = []any{}
+		}
+		var index int
+		if k, v, ok := extractKV(paths[1]); ok {
+			index = -1
+			for idx, cm := range base[seg].([]any) {
+				if MustCastAsMap(cm)[k] == v {
+					index = idx
+					break
+				}
+			}
+			if index == -1 {
+				return fmt.Errorf("element %v not found", paths[1])
+			}
+		} else if idx, ok := extractIndex(paths[1]); ok {
+			index = idx
+		} else {
+			return fmt.Errorf("unknown segment %v", paths[1])
+		}
+		l := base[seg].([]any)
+		if index < 0 || index >= len(l) {
+			// Index is greater, we need to append
+			if last {
+				l = append(l, value)
+			} else {
+				nm := Map{}
+				if err := setPathRecurse(nm, paths[2:], value); err != nil {
+					return err
+				}
+				l = append(l, nm)
+			}
+			base[seg] = l
+		} else {
+			v := MustCastAsMap(l[index])
+			if err := setPathRecurse(v, paths[2:], value); err != nil {
+				return err
+			}
+			l[index] = v
+		}
+	} else {
+		// This is a simple key traverse
+		// Find or create the target
+		// Create if needed
+		if _, f := base[seg]; !f {
+			base[seg] = map[string]any{}
+		}
+		if last {
+			base[seg] = value
+		} else {
+			return setPathRecurse(MustCastAsMap(base[seg]), paths[1:], value)
+		}
+	}
+	return nil
+}
+
+// GetValueForSetFlag parses the passed set flags which have format key=value and if any set the given path,
+// returns the corresponding value, otherwise returns the empty string. setFlags must have valid format.
+func GetValueForSetFlag(setFlags []string, path string) string {
+	ret := ""
+	for _, sf := range setFlags {
+		p, v := getPV(sf)
+		if p == path {
+			ret = v
+		}
+		// if set multiple times, return last set value
+	}
+	return ret
+}
+
+func fromJSON[T any](overlay []byte) (T, error) {
+	v := new(T)
+	err := json.Unmarshal(overlay, &v)
+	if err != nil {
+		return ptr.Empty[T](), err
+	}
+	return *v, nil
+}
+
+func tableLookup(v Map, simple string) (Map, bool) {
+	v2, ok := v[simple]
+	if !ok {
+		return nil, false
+	}
+	if vv, ok := v2.(map[string]interface{}); ok {
+		return vv, true
+	}
+
+	// This catches a case where a value is of type Values, but doesn't (for some
+	// reason) match the map[string]interface{}. This has been observed in the
+	// wild, and might be a result of a nil map of type Values.
+	if vv, ok := v2.(Map); ok {
+		return vv, true
+	}
+
+	return nil, false
+}
+
+func parsePath(key string) []string { return strings.Split(key, ".") }
+
+// alwaysString represents types that should always be decoded as strings
+// TODO: this could be automatically derived from the value_types.proto?
+var alwaysString = []string{
```

Comment:
Maybe `profile` as well. Right now, `--set profile=t` generates inconsistent result compared to others.